### PR TITLE
[FEATURE] Introduces view::get functionality, along with related exam…

### PR DESCRIPTION
**Pull request by joshuak94**
…ple snippets and test cases.

### CODE REVIEW

File: include/seqan3/range/view/get.hpp [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+ * \snippet test/snippet/range/view/get.cpp usage
+ * \hideinitializer
+ */
+template <size_t index>
+inline auto const get = view::transform([] (auto && in)
+{
+    using std::get;
+    // if constexpr (std::is_lvalue_reference_v<decltype(in)>)
+        // return static_cast<std::tuple_element_t<remove_cvref_t<decltype(in)>, index> &>(get<index>(std::forward<decltype(in)>(in)));
+    // else
```
Comment (marehr) [inline include/seqan3/range/view/get.hpp]: can we remove the commented out part?

File: include/seqan3/range/view/get.hpp [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+    // TODO: Currently does not return a reference.
```
Comment (marehr) [inline include/seqan3/range/view/get.hpp]: can you create an issue with an example what should work?

File: test/snippet/CMakeLists.txt [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+seqan3_require_test ()
+
+# add all snippets
+file(GLOB_RECURSE snippets "*.cpp")
+foreach(snippet ${snippets})
+    get_filename_component(target_cpp "${snippet}" NAME)
+    get_filename_component(snippet_full_path "${snippet}" DIRECTORY)
+
+    seqan3_snippet("${target_cpp}" "${snippet_full_path}")
+endforeach()
```
Comment (marehr) [inline test/snippet/CMakeLists.txt]: I hope this won't get in any conflict with your other prs :)

File: test/unit/range/view/view_get_test.cpp [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+
+    // combinability
+    dna4_vector cmp2{"TGCA"_dna4};
+    dna4_vector comp = qv | view::get<0> | view::complement;
+    EXPECT_EQ(cmp2, comp);
+
+    std::string cmp3{"TGCA"};
+    std::string to_char_test = comp | view::to_char;
+    EXPECT_EQ(cmp3, to_char_test);
+}
```
Comment (marehr) [inline test/unit/range/view/view_get_test.cpp]: can you add a test case for `std::tuple`, `std::pair`?

File: test/unit/range/view/view_get_test.cpp [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+    EXPECT_TRUE(std::ranges::InputRange<decltype(v1)>);
+    EXPECT_TRUE(std::ranges::ForwardRange<decltype(v1)>);
+    EXPECT_TRUE(std::ranges::BidirectionalRange<decltype(v1)>);
+    EXPECT_TRUE(std::ranges::RandomAccessRange<decltype(v1)>);
+    EXPECT_TRUE(std::ranges::View<decltype(v1)>);
+    EXPECT_TRUE(std::ranges::SizedRange<decltype(v1)>);
+    EXPECT_TRUE(std::ranges::CommonRange<decltype(v1)>);
+    EXPECT_TRUE(const_iterable_concept<decltype(v1)>);
+    EXPECT_FALSE((std::ranges::OutputRange<decltype(v1), std::tuple<int, int>>));
+    // EXPECT_TRUE((std::ranges::OutputRange<decltype(v1), int>));
```
Comment (marehr) [inline test/unit/range/view/view_get_test.cpp]: comment still needed?

**Comment by h-2** [comment]
This view definition exposes the reference on the returned view's element correctly
```cpp
template <size_t index>
inline auto const get = view::transform([] (auto && in)
    -> std::conditional_t<std::is_lvalue_reference_v<decltype(in)> && !std::is_const_v<decltype(in)>,
                          std::tuple_element_t<index, remove_cvref_t<decltype(in)>> &,
                          std::tuple_element_t<index, remove_cvref_t<decltype(in)>>>
{
    using std::get;
    return get<index>(in);
});
```

File: include/seqan3/range/view/get.hpp [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+ * |                                 |                                       |                                                    |
+ * | std::ranges::ViewableRange      | *required*                            | *preserved*                                        |
+ * | std::ranges::View               |                                       | *preserved*                                        |
+ * | std::ranges::SizedRange         |                                       | *preserved*                                        |
+ * | std::ranges::CommonRange        |                                       | *preserved*                                        |
+ * | std::ranges::OutputRange        |                                       | *preserved*                                        |
+ * | seqan3::const_iterable_concept  |                                       | *preserved*                                        |
+ * |                                 |                                       |                                                    |
+ * | seqan3::reference_t             | std::tuple                            |                                                    |
```
Comment (h-2) [inline include/seqan3/range/view/get.hpp]: change this line to:
` * | seqan3::reference_t             | seqan3::tuple_like_concept            | `decltype(get<index>(seqan3::reference_t<urng_t>{}))` | `

File: include/seqan3/range/view/get.hpp [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+ */
+template <size_t index>
+inline auto const get = view::transform([] (auto && in)
+{
+    using std::get;
+    // if constexpr (std::is_lvalue_reference_v<decltype(in)>)
+        // return static_cast<std::tuple_element_t<remove_cvref_t<decltype(in)>, index> &>(get<index>(std::forward<decltype(in)>(in)));
+    // else
+    // TODO: Currently does not return a reference.
+    return get<index>(std::forward<decltype(in)>(in));
```
Comment (h-2) [inline include/seqan3/range/view/get.hpp]: add a static assertion here that enforces the tuple_like_concept
```cpp
    static_assert(tuple_like_concept<decltype(in)>,
                  "You may only pass ranges to view::get whose reference_t models the tuple_like_concept.");
    return std::get<index>(in);
```
You need to add the include for `core/concept/tuple.hpp`.

For reasons unclear to me, you need to also adapt `core/concept/tuple.hpp` to include `core/pod_tuple.hpp` (while you are there you can remove the duplicate include-lines)..

File: include/seqan3/range/view/get.hpp [hunk 8]
```diff
@@ -0,0 +1,100 @@
...
+ */
+template <size_t index>
+inline auto const get = view::transform([] (auto && in)
+    -> std::conditional_t<std::is_lvalue_reference_v<decltype(in)> && !std::is_const_v<decltype(in)>,
+                          std::tuple_element_t<index, remove_cvref_t<decltype(in)>> &,
+                          std::tuple_element_t<index, remove_cvref_t<decltype(in)>>>
+{
+    using std::get;
+    return get<index>(std::forward<decltype(in)>(in));
+});
```
Comment (h-2) [inline include/seqan3/range/view/get.hpp]: add a static assertion here that enforces the tuple_like_concept
```cpp
    static_assert(tuple_like_concept<decltype(in)>,
                  "You may only pass ranges to view::get whose reference_t models the tuple_like_concept.");
    return std::get<index>(in);
```
You need to add the include for `core/concept/tuple.hpp`.

For reasons unclear to me, you need to also adapt `core/concept/tuple.hpp` to include `core/pod_tuple.hpp` (while you are there you can remove the duplicate include-lines)..

File: include/seqan3/range/view/get.hpp (hunk 6 above)
Comment (h-2) [reply include/seqan3/range/view/get.hpp]: `std::tuple_element_t<index, seqan3::reference_t<urng_t>>` may be nicer for the second column

File: include/seqan3/range/view/get.hpp [hunk 9]
```diff
@@ -0,0 +1,100 @@
...
+ *                      omitted in pipe notation]
+ * \tparam index        The index to get.
+ * \param[in] urange    The range being processed. [parameter is omitted in pipe notation]
+ * \returns             A range of std::get. See below for the properties of the returned range.
+ * \ingroup view
+ *
+ * ### View properties
+ *
+ * This view is a **deep view:** Given a range-of-tuples as input (as opposed to just a range), it will apply
+ * the transformation on the innermost tuple (instead of the outermost range).
```
Comment (h-2) [inline include/seqan3/range/view/get.hpp]: Oh, and this view is not a a deep view.

File: include/seqan3/range/view/get.hpp [hunk 10]
```diff
@@ -0,0 +1,100 @@
...
+/*!\name Alphabet related views
+ * \{
+ */
+
+/*!\brief               A view calling std::get on each element in a range.
+ * \tparam size_t       The type of the range being processed. See below for requirements. [template parameter is
+ *                      omitted in pipe notation]
+ * \tparam index        The index to get.
+ * \param[in] urange    The range being processed. [parameter is omitted in pipe notation]
+ * \returns             A range of std::get. See below for the properties of the returned range.
```
Comment (h-2) [inline include/seqan3/range/view/get.hpp]: "A range of elements where every element is the result of calling `std::get<index>` on the underlying element."

File: test/unit/range/view/view_get_test.cpp [hunk 11]
```diff
@@ -0,0 +1,183 @@
...
+    std::vector<dna4> cmprev2{dna4::T, dna4::G, dna4::C, dna4::A};
+    std::vector<dna4> revtest2 = t | view::get<0> | view::get<0> | view::reverse;
+    EXPECT_EQ(cmprev2, revtest2);
+
+    // reference check
+    functor0[0] = masked<dna4>{dna4::T, mask::UNMASKED};
+    std::vector<masked<dna4>> cmpref{{dna4::T, mask::UNMASKED}, {dna4::C, mask::UNMASKED},
+                                     {dna4::G, mask::MASKED}, {dna4::T, mask::UNMASKED}};
+    EXPECT_EQ(cmpref, functor0);
+}
```
Comment (h-2) [inline test/unit/range/view/view_get_test.cpp]: Good set of extensive tests :+1:

File: include/seqan3/range/view/get.hpp [hunk 12]
```diff
@@ -0,0 +1,101 @@
...
+ *
+ * \snippet test/snippet/range/view/get.cpp usage
+ * \hideinitializer
+ */
+template <size_t index>
+inline auto const get = view::transform([] (auto && in)
+    -> std::conditional_t<std::is_lvalue_reference_v<decltype(in)> && !std::is_const_v<decltype(in)>,
+                          std::tuple_element_t<index, remove_cvref_t<decltype(in)>> &,
+                          std::tuple_element_t<index, remove_cvref_t<decltype(in)>>>
+{
```
Comment (h-2) [inline include/seqan3/range/view/get.hpp]: Sorry, last comment, I checked this again and the correct approach is just to write:
```cpp
inline auto const get = view::transform([] (auto && in) -> auto &&
{
```
It also covers the `const`-cases and is much simpler to read.

Comment (joshuak94) [reply include/seqan3/range/view/get.hpp]: Done! Also rebased on master again.

**Comment by h-2** [comment]
The snippet test is still failing for this, can you rebase it on the current master branch and try again?

**Comment by joshuak94** [comment]
@h-2 done

**Pull request merged**